README
NAME
    Config::Pico - Pure Perl Interpretative COnfiguration = do "config.pl"

SYNOPSIS
        use Config::Pico;
    
        # general use
        @config = pico($basedir, "config.pl");
    
        # user agent
        $ua = LWP::UserAgent->new(pico "ua.pl");
    
        # logger
        $logger = Log::Dispatch->new(pico "logger.pl");
    
        # memcache
        $memd = Cache::Memcached->new(pico "memd.pl");
    
        # database
        $dbh = DBI->connect(pico "dbh.pl");

DESCRIPTION
    Use pure Perl to configure your app. This module exports "pico()"
    function which loads a configuration file written in Perl syntax. You
    don't have to learn any other languages/notations like YAML, XML, JSON,
    etc. to configure your Perl app as you already know Perl.

EXAMPLE CONFIGURATIONS
  ua.pl for LWP::UserAgent
        +(
            agent        => "Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 6.1; Trident/5.0)",
            max_redirect => 5,
            timeout      => 30,
            env_proxy    => 1,
        );

    This literary returns an array which represents four of key/value pair
    arguments to set up a new LWP::UserAgent instance. See LWP::UserAgent
    for more options.

  logger.pl for Log::Dispatch
        +(outputs => [
            ['File',   min_level => 'debug', filename => 'logfile'],
            ['Screen', min_level => 'warning'],
        ]);

    Yes, any Perl syntaxs, including "[]" arrayref etc., are available as
    it's a Perl script. See Log::Dispatch for more options.

  memd.pl for Cache::Memcached
        +{
            'servers' => [ "10.0.0.15:11211", "10.0.0.15:11212", "/var/sock/memcached",
                           "10.0.0.17:11211", [ "10.0.0.17:11211", 3 ] ],
            'debug' => 0,
            'compress_threshold' => 10_000,
        }

    This represents a hashref. See Cache::Memcached for more options.

  dbh.pl for DBD::SQLite
        +(
            "dbi:SQLite:dbname=dbfile", "", "", {
                AutoCommit        =>  1,
                RaiseError        =>  1,
                sqlite_unicode                   => 1,
                sqlite_allow_multiple_statements => 1,
            },
        );

    See DBI and DBD::SQLite for more options.

  dbh.pl for DBD::mysql
        local %_ = (
            driver    =>  'mysql',
            database  =>  'xxxx',
            host      =>  'localhost',
            port      =>  '3306',
            username  =>  'xxxxxx',
            password  =>  'xxxxxxxx',
            attr      =>  {
                AutoCommit        =>  1,
                RaiseError        =>  1,
                mysql_enable_utf8 =>  1,
            },
        );
    
        +("dbi:$_{driver}:database=$_{database};host=$_{host};port=$_{port}", $_{username}, $_{password}, $_{attr});

    Note that this uses a local variable which is easy for human to read.
    The last line represents an array packed for DBI. See DBD::mysql for
    more options.

FUNCTION
  pico([$dir,] $file)
    The first argument $dir is optional and specifies a base path. This
    would help you when you have variations in environments, e.g.
    development/staging/production etc. Put configuration files in a
    directory for each environments.

        my $conf = pico("config/$ENV{PLACK_ENV}", "dbi.pl");

    The last argument $file specifies a filename of configuration file
    written in Perl to load.

        my $conf = pico $file;

    In fact, this is equivalent to

        my $conf = do $file;

    The only difference between those two above is that, in a case that
    $file is not exist, "pico()" raises an error while "do()" doesn't. It'd
    be still easy for you to switch off the module and call "do()" instead
    of "pico()".

AUTHOR
    Yusuke Kawasaki http://www.kawa.net/

COPYRIGHT
    The following copyright notice applies to all the files provided in this
    distribution, including binary files, unless explicitly noted otherwise.

    Copyright 2012 Yusuke Kawasaki

LICENSE
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

SEE ALSO
    Config::PP - This provides "config_set()" function which dump a pure
    Perl file as well as "config_get()" function to load it like "pico()"
    does.



Changes
0.01  Tue Apr 24 15:09:11 PDT 2012
    - first shipping


t/config/array.pl
('foo', 'bar');


t/config/undef.pl
undef;


t/config/string.pl
'foobar';


make-dist.sh
#!/bin/sh

die () {
    echo "$*" >&2
    exit 1
}
doit () {
    echo "\$ $*" >&2
    $* || die "[ERROR:$?]"
}

[ -f Makefile ] && doit make clean
[ -f Makefile.old ] && doit /bin/rm -f Makefile.old
[ -f META.yml ] || doit touch META.yml

egrep -v '^(lib/.*\.pm|t/.*\.t)$' MANIFEST > MANIFEST~
ls Makefile.PL README Changes MANIFEST META.yml COPYING >> MANIFEST~ 2> /dev/null
find lib -type f -name '*.pm' >> MANIFEST~
ls t/*.t >> MANIFEST~
LC_ALL=C sort MANIFEST~ | uniq > MANIFEST~~
/bin/mv -f MANIFEST~~ MANIFEST~
diff MANIFEST MANIFEST~ > /dev/null || doit /bin/mv -f MANIFEST~ MANIFEST
/bin/rm -f MANIFEST~

doit perl Makefile.PL
doit make metafile
newmeta=`ls -t */META.yml | head -1`
diff META.yml $newmeta > /dev/null || doit /bin/cp -f $newmeta META.yml

doit make disttest

name=`grep '^name:' META.yml | sed 's#^.*: *##; s#-#/#g;'`
main=`grep "$name.pm$" < MANIFEST | head -1`
[ "$main" == "" ] && die "main module is not found in MANIFEST"
doit pod2text $main > README~
diff README README~ > /dev/null || doit /bin/mv -f README~ README
/bin/rm -f README~

doit make dist
[ -d blib ] && doit /bin/rm -fr blib
[ -f pm_to_blib ] && doit /bin/rm -f pm_to_blib
[ -f Makefile.old ] && doit /bin/rm -f Makefile.old

ls -lt *.tar.gz | head -1


lib/Config/Pico.pm
package Config::Pico;

use strict;
use warnings;
our $VERSION = '0.01';

use base 'Exporter';
use Carp ();
use File::Spec ();

our @EXPORT = qw(pico);

sub pico {
    my $base = $ENV{PICO_BASE} || "";
    $base = shift if (@_ > 1);
    my $file = shift;
    Carp::croak "'pico([\$base,] \$file)': too many arguments" if @_;
    $file = File::Spec->catfile($base, $file) if $base;
    Carp::croak "'$file': $!" unless -f $file;
    my @conf = do $file;
    if (@conf == 1 && ! defined $conf[0]) {
        Carp::croak "'$file': $@" if $@;
        Carp::croak "'$file': $!" if $!;
    }
    wantarray ? @conf : pop @conf;
}

1;
__END__

=encoding utf-8

=for stopwords

=head1 NAME

Config::Pico - Perl Interpretative COnfiguration: do "config.pl"

=head1 SYNOPSIS

    use Config::Pico;
    
    # general use
    @config = pico($basedir, "config.pl");

    # user agent
    $ua = LWP::UserAgent->new(pico "lwp.pl");

    # logger
    $logger = Log::Dispatch->new(pico "log.pl");

    # database
    $dbh = DBI->connect(pico "dbi.pl");

lwp.pl configuration for L<LWP::UserAgent>

    +(
        agent        => "Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 6.1; Trident/5.0)",
        max_redirect => 5,
        timeout      => 30,
        env_proxy    => 1,
    ); 

log.pl configuration for L<Log::Dispatch>

    +(outputs => [
        ['File',   min_level => 'debug', filename => 'logfile'],
        ['Screen', min_level => 'warning'],
    ]);

dbi.pl configuration for L<DBI>, L<DBD::SQLite>

    +(
        "dbi:SQLite:dbname=dbfile", "", "", {
            AutoCommit        =>  1,
            RaiseError        =>  1,
            sqlite_unicode                   => 1,
            sqlite_allow_multiple_statements => 1,
        },
    );

another configuration for L<DBD::MySQL>

    local %_ = (
        driver    =>  'mysql',
        database  =>  'xxxx',
        host      =>  'localhost',
        port      =>  '3306',
        username  =>  'xxxxxx',
        password  =>  'xxxxxxxx',
        attr      =>  {
            AutoCommit        =>  1,
            RaiseError        =>  1,
	    mysql_enable_utf8 =>  1,
        },
    );
    
    +("dbi:$_{driver}:database=$_{database};host=$_{host};port=$_{port}", $_{username}, $_{password}, $_{attr});

=head1 DESCRIPTION

Use Perl as a DSL to configure you app. This module exports C<pico()>
function which loads a configuration file written in Perl syntax. 
You don't have to learn other languages/notations like YAML, XML,
JSON, etc. as you already know Perl.

=head1 FUNCTION

=head2 pico([$dir,] $file)

The first argument C<$dir> is optional and specifies a base path.
This would help you when you have variations in environments
which need a set of configuration files respectively.

    my $conf = pico("config/$ENV{PLACK_ENV}", "dbi.pl");

The last argument C<$file> specifies a filename of configuration
written in Perl to load.

    my $conf = pico $file;

is equivalent to

    my $conf = do $file;

Another difference between those two is that, in case that C<$file>
is not exist, C<pico()> raises an error while C<do()> doesn't.
It'd be still easy for you to switch off the module and call C<do()>
instead of C<pico()>.

=head1 AUTHOR

Yusuke Kawasaki http://www.kawa.net/

=head1 COPYRIGHT

The following copyright notice applies to all the files provided in
this distribution, including binary files, unless explicitly noted
otherwise.

Copyright 2012 Yusuke Kawasaki

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
